From e1535070d52e75d4393459b8a904729a664155f2 Mon Sep 17 00:00:00 2001
From: "silipwn [Ashwin]" <nambiar1@purdue.edu>
Date: Wed, 7 Feb 2024 19:14:12 +0000
Subject: [PATCH] Add support for auto spawning on remote sessions

---
 Tools/autotest/sim_vehicle.py | 51 ++++++++++++++++++++++-------------
 1 file changed, 33 insertions(+), 18 deletions(-)

diff --git a/Tools/autotest/sim_vehicle.py b/Tools/autotest/sim_vehicle.py
index 3f80bbd181..9c548aa6ed 100755
--- a/Tools/autotest/sim_vehicle.py
+++ b/Tools/autotest/sim_vehicle.py
@@ -601,28 +601,43 @@ def run_in_terminal_window(name, cmd, **kw):
     runme.extend(cmd)
     progress_cmd("Run " + name, runme)
 
-    if under_macos() and os.environ.get('DISPLAY'):
-        # on MacOS record the window IDs so we can close them later
-        out = subprocess.Popen(runme, stdout=subprocess.PIPE, **kw).communicate()[0]
-        out = out.decode('utf-8')
-        p = re.compile('tab 1 of window id (.*)')
+    # Check if we are inside a tmux session
+    if os.environ.get('TMUX') is not None:
+        progress("Detected TMUX session; Spawning sims in a new window")
+        # Save the current window number
+        window_number = subprocess.check_output(['tmux', 'display-message', '-p', '#I']).decode('utf-8').strip()
+        # Create a window name with prefix sim_vehicle_spawn and then just use a random car
+        window_name = "sim_vehicle_spawn-" + str(int(time.time()))
+        # Run the command in a new TMUX window
+        subprocess.run(['tmux', 'new-window', '-n', window_name])
+        subprocess.run(['tmux', 'send-keys'," ".join(cmd), 'Enter'])
+        # Come back to the original window
+        subprocess.run(['tmux', 'select-window', '-t', window_number])
+
+    # Check if we are inside GUI & Behave normally
+    else:
+        if under_macos() and os.environ.get('DISPLAY'):
+            # on MacOS record the window IDs so we can close them later
+            out = subprocess.Popen(runme, stdout=subprocess.PIPE, **kw).communicate()[0]
+            out = out.decode('utf-8')
+            p = re.compile('tab 1 of window id (.*)')
 
-        tstart = time.time()
-        while time.time() - tstart < 5:
-            tabs = p.findall(out)
+            tstart = time.time()
+            while time.time() - tstart < 5:
+                tabs = p.findall(out)
 
-            if len(tabs) > 0:
-                break
+                if len(tabs) > 0:
+                    break
 
-            time.sleep(0.1)
-        # sleep for extra 2 seconds for application to start
-        time.sleep(2)
-        if len(tabs) > 0:
-            windowID.append(tabs[0])
+                time.sleep(0.1)
+            # sleep for extra 2 seconds for application to start
+            time.sleep(2)
+            if len(tabs) > 0:
+                windowID.append(tabs[0])
+            else:
+                progress("Cannot find %s process terminal" % name)
         else:
-            progress("Cannot find %s process terminal" % name)
-    else:
-        subprocess.Popen(runme, **kw)
+            subprocess.Popen(runme, **kw)
 
 
 tracker_uarta = None  # blemish
-- 
2.34.1

